core/src/main/java/com/linecorp/armeria/internal/common/DefaultCancellationScheduler.java

```diff
@@ -76,295 +58,233 @@ final class DefaultCancellationScheduler implements CancellationScheduler {
     private long startTimeNanos;
     @Nullable
     private EventExecutor eventLoop;
-    @Nullable
-    private CancellationTask task;
-    @Nullable
-    private volatile Runnable pendingTask;
+    private volatile CancellationTask task = noopCancellationTask;
     @Nullable
     private ScheduledFuture<?> scheduledFuture;
+    private long setFromNowStartNanos;
+    private TimeoutMode timeoutMode = TimeoutMode.SET_FROM_START;
+    @Nullable
+    private volatile Throwable cause;
+    private final Ticker ticker;
+    private final ReentrantShortLock lock = new ReentrantShortLock();
+
+    private final boolean server;
     @Nullable
     private volatile CancellationFuture whenCancelling;
     @Nullable
     private volatile CancellationFuture whenCancelled;
-    @Nullable
-    private volatile TimeoutFuture whenTimingOut;
-    @Nullable
-    private volatile TimeoutFuture whenTimedOut;
-    @SuppressWarnings("FieldMayBeFinal")
-    private volatile long pendingTimeoutNanos;
-    private final boolean server;
-    @Nullable
-    private Throwable cause;
 
     @VisibleForTesting
     DefaultCancellationScheduler(long timeoutNanos) {
         this(timeoutNanos, true);
     }
 
     DefaultCancellationScheduler(long timeoutNanos, boolean server) {
+        this(timeoutNanos, server, Ticker.systemTicker());
+    }
+
+    @VisibleForTesting
+    DefaultCancellationScheduler(long timeoutNanos, boolean server, Ticker ticker) {
         this.timeoutNanos = timeoutNanos;
-        pendingTimeoutNanos = timeoutNanos;
         this.server = server;
+        this.ticker = ticker;
     }
 
     /**
      * Initializes this {@link DefaultCancellationScheduler}.
      */
     @Override
     public void initAndStart(EventExecutor eventLoop, CancellationTask task) {
```

Comment:
Q) Would it be possible allocating a random `EventLoop` in the constructor? If we decided to use `lock`, I don't see we should use the same event loop of a request for `CancellationScheduler`.

Reply:
I imagine usually the event loop will time out within the `HttpResponseWrapper` or `AbstractHttpRequestHandler`. If the event loop for the `CancellationScheduler` is different from the channel event loop, I think there can be unnecessary event loop rescheduling.

Is there a reason you prefer assigning a random event loop?

Reply:
The thing is, `ClientRequestContext` relies on `CancellationScheduler` for certain APIs (e.g. `ClientRequestContext#responseTimeoutMillis`). `ClientRequestContext` is assigned an `EventExecutor` after initialization, so `CancellationScheduler` is following this pattern by assigning an `EventExecutor` later on.

Perhaps if the relationship between `EventExecutor` and `ClientRequestContext` is modified, `CancellationScheduler` can be simpler. Obviously, this is out of this PR's scope though

Reply:
It was just a random idea but I am satisfied with the current implementation.